File: benchmarks/references/ground-truth/R/tc-002-demographics.R
#!/usr/bin/env Rscript
# TC-002 Ground Truth: Baseline Demographics Table
# Part of the ASA Biopharm AI/ML WG Agentic AI Benchmark
#
# Cross-validated against: SAS PROC FREQ + PROC MEANS, Python pandas + gtsummary
# Dependencies: dplyr, jsonlite, Tplyr (or tableone)
# Usage: Rscript tc-002-demographics.R --seed 42 --n 400 --output results.json

library(dplyr)
library(jsonlite)

# Source common utilities
source("common/data-generation.R")

# ─────────────────────────────────────────────────────
# Command-line argument parsing
# ─────────────────────────────────────────────────────

parse_args <- function() {
  args <- commandArgs(trailingOnly = TRUE)

  get_arg <- function(flag, default) {
    idx <- which(args == flag)
    if (length(idx) > 0 && idx < length(args)) {
      return(args[idx + 1])
    }
    return(default)
  }

  list(
    seed = as.integer(get_arg("--seed", "42")),
    n = as.integer(get_arg("--n", "400")),
    data = get_arg("--data", NA),
    output = get_arg("--output", NA)
  )
}

# ─────────────────────────────────────────────────────
# TC-002 Core Computation
# ─────────────────────────────────────────────────────

compute_demographics <- function(adsl, population = "SAFETY", seed = NA) {

  # Apply population filter
  if (population == "SAFETY") {
    data <- adsl %>% filter(SAFFL == "Y")
  } else if (population == "ITT") {
    data <- adsl %>% filter(ITTFL == "Y")
  } else {
    data <- adsl
  }

  if (nrow(data) == 0) {
    stop("No subjects in the specified population")
  }

  n_total <- nrow(data)

  # ── Continuous: AGE ──
  age_stats <- data %>%
    group_by(TRT01PN, TRT01P) %>%
    summarise(
      n = n(),
      mean = mean(AGE, na.rm = TRUE),
      sd = sd(AGE, na.rm = TRUE),
      median = median(AGE, na.rm = TRUE),
      min = min(AGE, na.rm = TRUE),
      max = max(AGE, na.rm = TRUE),
      .groups = "drop"
    )

  # ── Categorical: SEX, RACE, REGION1, ECOG ──
  compute_freq <- function(data, var, var_name) {
    data %>%
      group_by(TRT01PN, TRT01P, !!sym(var)) %>%
      summarise(n = n(), .groups = "drop") %>%
      group_by(TRT01PN, TRT01P) %>%
      mutate(pct = round(100 * n / sum(n), 1)) %>%
      ungroup() %>%
      mutate(variable = var_name) %>%
      rename(level = !!sym(var)) %>%
      # BUGFIX: coerce level to character so bind_rows() can combine numeric
      # (ECOG) and character (SEX/RACE/REGION) categorical levels without a
      # vctrs type error. This previously crashed TC-002 in R entirely.
      mutate(level = as.character(level))
  }

  sex_tbl   <- compute_freq(data, "SEX", "Sex")
  race_tbl  <- compute_freq(data, "RACE", "Race")
  region_tbl <- compute_freq(data, "REGION1", "Region")
  ecog_tbl  <- compute_freq(data, "ECOG", "ECOG")

  cat_tbl <- bind_rows(sex_tbl, race_tbl, region_tbl, ecog_tbl)

  # ── Overall totals ──
  total_age <- data %>%
    summarise(
      n = n(),
      mean = mean(AGE, na.rm = TRUE),
      sd = sd(AGE, na.rm = TRUE),
      median = median(AGE, na.rm = TRUE),
      min = min(AGE, na.rm = TRUE),
      max = max(AGE, na.rm = TRUE)
    )

  # ── Build result ──
  result <- list(
    test_case_id = "TC-002",
    variant_id = if (is.na(seed)) NA else paste0("v", seed),
    language = "R",
    package = "dplyr",
    package_version = as.character(packageVersion("dplyr")),
    population = population,
    n_total = n_total,
    age_by_arm = as.data.frame(age_stats) %>%
      mutate(across(where(is.numeric), ~ round(., 2))),
    categorical_by_arm = as.data.frame(cat_tbl) %>%
      mutate(across(where(is.numeric), ~ round(., 2))),
    total_age = as.data.frame(total_age) %>%
      mutate(across(where(is.numeric), ~ round(., 2))),
    seed = seed
  )

  return(result)
}

# ─────────────────────────────────────────────────────
# Main
# ─────────────────────────────────────────────────────

if (sys.nframe() == 0) {
  opts <- parse_args()
  seed <- opts$seed
  n <- opts$n

  cat(sprintf("TC-002: Baseline Demographics (R) — seed=%d, n=%d\n", seed, n))

  # Obtain ADSL: shared CSV (cross-language) or in-language generation (smoke)
  adsl <- get_adsl(data_path = opts$data, seed = seed, n_subjects = n, n_arms = 2)
  cat(sprintf("%s ADSL with %d subjects\n",
              if (!is.na(opts$data)) "Loaded shared" else "Generated", nrow(adsl)))

  # Compute demographics
  result <- compute_demographics(adsl, population = "SAFETY", seed = seed)

  # Print summary
  cat("\n──────────────────────────────────────────────\n")
  cat(sprintf("Population: %s (N=%d)\n", result$population, result$n_total))

  cat("\n— Age by Arm —\n")
  print(result$age_by_arm)

  cat("\n— Sex by Arm —\n")
  print(result$categorical_by_arm %>% filter(variable == "Sex"))

  cat("\n— Race by Arm —\n")
  print(result$categorical_by_arm %>% filter(variable == "Race"))

  cat("\n— Region by Arm —\n")
  print(result$categorical_by_arm %>% filter(variable == "Region"))

  cat("\n— ECOG by Arm —\n")
  print(result$categorical_by_arm %>% filter(variable == "ECOG"))

  cat("──────────────────────────────────────────────\n")

  # Structured JSON output
  print_output(result)

  if (!is.na(opts$output)) {
    write_output(result, opts$output)
  }
}

File: benchmarks/references/ground-truth/Python/tc_002_demographics.py
#!/usr/bin/env python3
"""TC-002 Ground Truth: Baseline Demographics Table (Python)

Part of the ASA Biopharm AI/ML WG Agentic AI Benchmark.

Cross-validated against: R dplyr + Tplyr, SAS PROC FREQ + PROC MEANS
Dependencies: pandas, numpy
Usage: python tc_002_demographics.py --seed 42 --n 400 --output results.json
"""

import argparse
import json
import sys
from pathlib import Path

import numpy as np
import pandas as pd

from common.data_generation import get_adsl


def compute_demographics(adsl: pd.DataFrame, population: str = "SAFETY",
                         seed: int = None) -> dict:
    """Compute baseline demographics summary by treatment arm.

    Args:
        adsl: ADSL-like DataFrame
        population: Population filter ("SAFETY" or "ITT")

    Returns:
        dict with structured benchmark output
    """
    # Population filter
    if population == "SAFETY":
        data = adsl[adsl["SAFFL"] == "Y"].copy()
    elif population == "ITT":
        data = adsl[adsl["ITTFL"] == "Y"].copy()
    else:
        data = adsl.copy()

    n_total = len(data)

    if n_total == 0:
        raise ValueError("No subjects in the specified population")

    # ── Continuous: AGE ──
    age_stats = (
        data.groupby(["TRT01PN", "TRT01P"])["AGE"]
        .agg(["count", "mean", "std", "median", "min", "max"])
        .round(2)
        .reset_index()
    )

    # ── Categorical variables ──
    def compute_cat_freq(df, var, var_name):
        """Compute frequency table for a categorical variable by arm."""
        freq = (
            df.groupby(["TRT01PN", "TRT01P", var])
            .size()
            .reset_index(name="n")
        )
        freq["pct"] = (
            freq.groupby(["TRT01PN", "TRT01P"])["n"]
            .transform(lambda x: round(100 * x / x.sum(), 1))
        )
        freq["variable"] = var_name
        freq = freq.rename(columns={var: "level"})
        return freq

    cat_vars = {
        "SEX": "Sex",
        "RACE": "Race",
        "REGION1": "Region",
        "ECOG": "ECOG",
    }

    cat_tables = []
    for var, name in cat_vars.items():
        if var in data.columns:
            cat_tables.append(compute_cat_freq(data, var, name))

    cat_all = pd.concat(cat_tables, ignore_index=True)

    # ── Overall totals ──
    total_age = data["AGE"].agg(["count", "mean", "std", "median", "min", "max"]).round(2)

    # ── Build result ──
    result = {
        "test_case_id": "TC-002",
        "variant_id": (f"v{seed}" if seed is not None else None),
        "language": "Python",
        "package": "pandas",
        "package_version": pd.__version__,
        "population": population,
        "n_total": n_total,
        "age_by_arm": age_stats.to_dict(orient="records"),
        "categorical_by_arm": cat_all.to_dict(orient="records"),
        "total_age": total_age.to_dict(),
        "seed": seed,
    }

    return result


def main():
    parser = argparse.ArgumentParser(
        description="TC-002: Baseline Demographics Table (Python)"
    )
    parser.add_argument("--seed", type=int, default=42)
    parser.add_argument("--n", type=int, default=400)
    parser.add_argument("--data", type=str, default=None,
                        help="Shared ADSL CSV (for cross-language verification)")
    parser.add_argument("--output", type=str, default=None)
    args = parser.parse_args()

    print(f"TC-002: Baseline Demographics (Python) — seed={args.seed}, n={args.n}")

    # Obtain ADSL: shared CSV (cross-language) or in-language generation (smoke)
    adsl = get_adsl(data_path=args.data, seed=args.seed, n_subjects=args.n, n_arms=2)
    print(f"{'Loaded shared' if args.data else 'Generated'} ADSL with {len(adsl)} subjects")

    # Compute demographics
    result = compute_demographics(adsl, population="SAFETY", seed=args.seed)

    # Print summary
    print("\n" + "─" * 50)
    print(f"Population: {result['population']} (N={result['n_total']})")
    print("\n— Age by Arm —")
    for row in result["age_by_arm"]:
        print(f"  Arm {row['TRT01PN']}: n={row['count']}, "
              f"mean={row['mean']}, sd={row['std']}, "
              f"median={row['median']}, range=({row['min']}, {row['max']})")

    print("\n— Categorical by Arm —")
    for row in result["categorical_by_arm"]:
        print(f"  {row['variable']}: {row['level']} (Arm {row['TRT01PN']}): "
              f"{row['n']} ({row['pct']}%)")
    print("─" * 50)

    # Structured JSON
    print("\n=== BENCHMARK OUTPUT ===")
    print(json.dumps(result, indent=2, default=str))
    print("=== END OUTPUT ===")

    if args.output:
        outpath = Path(args.output)
        outpath.parent.mkdir(parents=True, exist_ok=True)
        with open(outpath, "w") as f:
            json.dump(result, f, indent=2, default=str)
        print(f"Wrote output to: {outpath}")


if __name__ == "__main__":
    main()
